py/desispec/io/spectra.py

```diff
@@ -241,63 +271,123 @@ def read_spectra(infile, single=False):
         if name == "FIBERMAP":
-            fmap = encode_table(Table(hdus[h].read(), copy=True).as_array())
+            if not skip_hdu["FIBERMAP"]:
+                fmap = encode_table(
+                    Table(
+                        hdus[h].read(rows=rows, columns=select_columns["FIBERMAP"]),
+                        copy=True,
+                    ).as_array()
+                )
         elif name == "EXP_FIBERMAP":
-            expfmap = encode_table(Table(hdus[h].read(), copy=True).as_array())
+            if not skip_hdu["EXP_FIBERMAP"]:
+                expfmap = encode_table(
+                    Table(
+                        hdus[h].read(rows=rows, columns=select_columns["EXP_FIBERMAP"]),
+                        copy=True,
+                    ).as_array()
+                )
         elif name == "SCORES":
-            scores = encode_table(Table(hdus[h].read(), copy=True).as_array())
-        elif name == 'EXTRA_CATALOG':
-            extra_catalog = encode_table(Table(hdus[h].read(), copy=True).as_array())
+            if not skip_hdu["SCORES"]:
+                scores = encode_table(
+                    Table(
+                        hdus[h].read(rows=rows, columns=select_columns["SCORES"]),
+                        copy=True,
+                    ).as_array()
+                )
+        elif name == "EXTRA_CATALOG":
+            if not skip_hdu["EXTRA_CATALOG"]:
+                extra_catalog = encode_table(
+                    Table(
+                        hdus[h].read(
+                            rows=rows, columns=select_columns["EXTRA_CATALOG"]
+                        ),
+                        copy=True,
+                    ).as_array()
+                )
         else:
             # Find the band based on the name
             mat = re.match(r"(.*)_(.*)", name)
             if mat is None:
-                raise RuntimeError("FITS extension name {} does not contain the band".format(name))
+                raise RuntimeError(
+                    "FITS extension name {} does not contain the band".format(name)
+                )
             band = mat.group(1).lower()
             type = mat.group(2)
             if band not in bands:
                 bands.append(band)
             if type == "WAVELENGTH":
                 if wave is None:
                     wave = {}
-                #- Note: keep original float64 resolution for wavelength
+                # - Note: keep original float64 resolution for wavelength
                 wave[band] = native_endian(hdus[h].read())
             elif type == "FLUX":
                 if flux is None:
                     flux = {}
-                flux[band] = native_endian(hdus[h].read().astype(ftype))
+                if np.any(rows == None):
+                    flux[band] = native_endian(ahdus[h].data.astype(ftype))
+                else:
+                    flux[band] = native_endian(ahdus[h].section[rows, :].astype(ftype))
```

Comment:
Since you have to if/else the slicing of the 2D arrays anyway, let's use the fitsio `hdus` when not slicing instead of the astropy `ahdus`.  For non-compressed files it doesn't make a big difference, but for compressed files (like the non-coadded spectra*.fits.gz files) it can be a big difference.  i.e.
```python
                if rows is None:
                    flux[band] = native_endian(hdus[h].read().astype(ftype))
                else:
                    flux[band] = native_endian(ahdus[h].section[rows, :].astype(ftype))
```

Reply:
updated to use fitsio `hdus` whenever possible for efficiency.